hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataCompactResolver.java

```diff
@@ -102,10 +112,56 @@ public KvMetadata resolveMetadata(
         }
         maybeAddDefaultField(isKey, resolvedFields, fields, QueryDataType.OBJECT);
 
+        Schema schema = resolveSchema(typeName, fieldsByPath);
+
         return new KvMetadata(
                 fields,
                 GenericQueryTargetDescriptor.DEFAULT,
-                new CompactUpsertTargetDescriptor(typeName)
+                new CompactUpsertTargetDescriptor(schema)
         );
     }
+
+    private Schema resolveSchema(String typeName, Map<QueryPath, MappingField> fields) {
+        SchemaWriter schemaWriter = new SchemaWriter(typeName);
+        for (Entry<QueryPath, MappingField> entry : fields.entrySet()) {
+            String name = entry.getKey().getPath();
+            QueryDataType type = entry.getValue().type();
+            schemaWriter.addField(new FieldDescriptor(name, resolveToCompactType(type.getTypeFamily())));
+        }
+        return schemaWriter.build();
+    }
+
+    @SuppressWarnings("checkstyle:ReturnCount")
+    private static FieldType resolveToCompactType(QueryDataTypeFamily type) {
+        switch (type) {
+            case BOOLEAN:
+                return FieldType.BOOLEAN;
+            case TINYINT:
+                return FieldType.BYTE;
+            case SMALLINT:
+                return FieldType.SHORT;
+            case INTEGER:
+                return FieldType.INT;
+            case BIGINT:
+                return FieldType.LONG;
+            case REAL:
+                return FieldType.FLOAT;
+            case DOUBLE:
+                return FieldType.DOUBLE;
+            case DECIMAL:
+                return FieldType.DECIMAL;
+            case VARCHAR:
+                return FieldType.UTF;
+            case TIME:
+                return FieldType.TIME;
+            case DATE:
+                return FieldType.DATE;
+            case TIMESTAMP:
+                return FieldType.TIMESTAMP;
+            case TIMESTAMP_WITH_TIME_ZONE:
+                return FieldType.TIMESTAMP_WITH_TIMEZONE;
+            default:
+                throw QueryException.error("Compact format does not allow " + type + " data type");
```

Comment:
We should never reach this point - maybe instead of `QueryException` `IllegalArgumentException` should be thrown.